File: cqf-fhir-utility/src/test/java/org/opencds/cqf/fhir/utility/operation/MethodBinderTest.java
package org.opencds.cqf.fhir.utility.operation;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import ca.uhn.fhir.model.api.annotation.Description;
import ca.uhn.fhir.rest.annotation.IdParam;
import ca.uhn.fhir.rest.annotation.Operation;
import ca.uhn.fhir.rest.annotation.OperationParam;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import org.hl7.fhir.dstu3.model.Measure;
import org.hl7.fhir.instance.model.api.IBaseParameters;
import org.hl7.fhir.instance.model.api.IBaseResource;
import org.hl7.fhir.r4.model.IdType;
import org.hl7.fhir.r4.model.Parameters;
import org.hl7.fhir.r4.model.StringType;
import org.junit.jupiter.api.Test;

public class MethodBinderTest {

    // These are simply a bunch of example method signatures to test the MethodBinder class
    class ExampleMethods {

        public void noAnnotation() {}

        @Operation(name = "idFirst")
        public void idFirst(@IdParam IdType id) {}

        @Operation(name = "idLast")
        public void idLast(@OperationParam(name = "param") StringType param, @IdParam IdType id) {}

        @Operation(name = "noId")
        public void noId(@OperationParam(name = "param") StringType param) {}

        @Operation(name = "manyIds")
        public void manyIds(@IdParam IdType id, @IdParam IdType id2) {}

        @Operation(name = "idIsNotIdType")
        public void idIsNotIdType(@IdParam StringType id) {}

        @Operation(name = "extraFirst")
        public void extraFirst(@ExtraParams IBaseParameters extras, @OperationParam(name = "param") String param) {}

        @Operation(name = "extraLast")
        public void extraLast(
                @OperationParam(name = "param") StringType param, @ExtraParams IBaseParameters parameters) {}

        @Operation(name = "manyExtra")
        public void manyExtra(@ExtraParams IBaseParameters parameters, @ExtraParams IBaseParameters parameters2) {}

        @Operation(name = "extraIsNotParametersType")
        public void extraIsNotParametersType(@ExtraParams IdType idType) {}

        @Operation(name = "operationParamIsNotFhirType")
        public void operationParamIsNotFhirType(@OperationParam(name = "stringType") String stringType) {}

        @Description("has a description")
        @Operation(name = "hasDescription")
        public void hasDescription(@IdParam IdType id) {}

        @Operation(name = "hasCanonical", canonicalUrl = "http://example.com")
        public void hasCanonical(@IdParam IdType id) {}

        @Operation(name = "hasTypeName", typeName = "Measure")
        public void hasTypeName(@OperationParam(name = "stringType") StringType stringType) {}

        @Operation(name = "hasType", type = Measure.class)
        public void hasType(@OperationParam(name = "stringType") StringType stringType) {}

        @Operation(name = "noArgs")
        public void noArgs() {}

        @Operation(name = "$nameNormalizes")
        public void nameNormalizes() {}
    }

    private static List<Method> methods = Arrays.asList(ExampleMethods.class.getDeclaredMethods());

    private static Method methodByName(String name) {
        return methods.stream()
                .filter(x -> x.getName().equals(name))
                .findFirst()
                .get();
    }

    private static MethodBinder methodBinderByName(String name) {
        return new MethodBinder(methodByName(name));
    }

    @Test
    public void nullValue_throws() {
        assertThrows(NullPointerException.class, () -> new MethodBinder(null));
    }

    @Test
    public void noAnnotation_throws() {
        assertThrows(NullPointerException.class, () -> new MethodBinder(methodByName("noAnnotation")));
    }

    @Test
    public void idFirst() {
        var m = methodBinderByName("idFirst");
        assertEquals(Scope.INSTANCE, m.scope());
        assertEquals("idFirst", m.name());
        assertEquals(1, m.parameters().size());
    }

    @Test
    public void idLast_throws() {
        assertThrows(IllegalArgumentException.class, () -> methodBinderByName("idLast"));
    }

    @Test
    public void manyIds_throws() {
        assertThrows(IllegalArgumentException.class, () -> methodBinderByName("manyIds"));
    }

    @Test
    public void idParamNotIdType_throws() {
        assertThrows(IllegalArgumentException.class, () -> methodBinderByName("idIsNotIdType"));
    }

    @Test
    public void noId() {
        var m = methodBinderByName("noId");
        assertEquals(Scope.SERVER, m.scope());
        assertEquals("noId", m.name());
        assertEquals(1, m.parameters().size());
    }

    @Test
    public void extraFirst_throws() {
        assertThrows(IllegalArgumentException.class, () -> methodBinderByName("extraFirst"));
    }

    @Test
    public void extraLast() {
        var m = methodBinderByName("extraLast");
        assertEquals(Scope.SERVER, m.scope());
        assertEquals("extraLast", m.name());
        assertEquals(2, m.parameters().size());
    }

    @Test
    public void manyExtra_throws() {
        assertThrows(IllegalArgumentException.class, () -> methodBinderByName("manyExtra"));
    }

    @Test
    public void extraIsNotParametersType_throws() {
        assertThrows(IllegalArgumentException.class, () -> methodBinderByName("extraIsNotParametersType"));
    }

    @Test
    public void operationParamIsNotFhirType_throws() {
        assertThrows(IllegalArgumentException.class, () -> methodBinderByName("operationParamIsNotFhirType"));
    }

    @Test
    public void binderGetsDescription() {
        var m = methodBinderByName("hasDescription");
        assertEquals("has a description", m.description());
    }

    @Test
    public void binderGetsCanonical() {
        var m = methodBinderByName("hasCanonical");
        assertEquals("http://example.com", m.canonicalUrl());
    }

    @Test
    public void binderGetsMethod() {
        var m = methodBinderByName("idFirst");
        assertNotNull(m.method());
        assertEquals(m.method().getName(), "idFirst");
    }

    @Test
    public void binderGetsOperationAnnotation() {
        var m = methodBinderByName("idFirst");
        assertNotNull(m.operation());
        assertEquals(m.operation().name(), "idFirst");
    }

    @Test
    public void hasType() {
        var m = methodBinderByName("hasType");
        assertEquals("Measure", m.typeName());
        assertEquals(Scope.TYPE, m.scope());
    }

    @Test
    public void hasTypeName() {
        var m = methodBinderByName("hasTypeName");
        assertEquals("Measure", m.typeName());
        assertEquals(Scope.TYPE, m.scope());
    }

    @Test
    public void noArgs() {
        var m = methodBinderByName("noArgs");
        assertEquals(Scope.SERVER, m.scope());
    }

    @Test
    public void nameNormalizes() {
        var m = methodBinderByName("nameNormalizes");
        assertEquals("nameNormalizes", m.name());
    }

    @Test
    public void unboundArgs_throws() {
        var m = methodBinderByName("noArgs");

        var provider = new ExampleMethods();
        var parameters = new Parameters().addParameter("anyParam", new StringType("anyValue"));
        var e = assertThrows(IllegalArgumentException.class, () -> m.bind(provider, null, parameters));
        assertTrue(e.getMessage().contains("not bound"));
    }

    @Test
    public void idWithoutInstanceScope_throws() {
        var m = methodBinderByName("noArgs");

        var provider = new ExampleMethods();
        var id = new IdType("Library/123");
        var e = assertThrows(IllegalArgumentException.class, () -> m.bind(provider, id, null));
        assertTrue(e.getMessage().contains("non-instance"));
    }

    @Test
    public void instanceScopeWithoutId_throws() {
        var m = methodBinderByName("idFirst");

        var provider = new ExampleMethods();
        var e = assertThrows(IllegalArgumentException.class, () -> m.bind(provider, null, null));
        assertTrue(e.getMessage().contains("id required"));
    }

    @Test
    public void missingOperationParamAnnotation_throws() {
        final class MissingOperationAnnotation {
            @Operation(name = "missingOperationParam")
            public IBaseResource missingOperationParam(StringType param) {
                return null;
            }
        }

        var method = MissingOperationAnnotation.class.getDeclaredMethods()[0];
        var e = assertThrows(IllegalArgumentException.class, () -> new MethodBinder(method));
        assertTrue(e.getMessage().contains("must be annotated"));
    }

    @Test
    public void conflictingAnnotations_throws() {
        final class ConflictingAnnotation {
            @Operation(name = "function")
            public IBaseResource function(@IdParam @OperationParam(name = "id") IdType id) {
                return null;
            }
        }

        var method = ConflictingAnnotation.class.getDeclaredMethods()[0];
        var e = assertThrows(IllegalArgumentException.class, () -> new MethodBinder(method));
        assertTrue(e.getMessage().contains("one of"));
    }

    @Test
    public void missingOperationAnnotation_throws() {
        final class MissingOperationAnnotation {
            @SuppressWarnings("unused")
            public IBaseResource function(@IdParam IdType id) {
                return null;
            }
        }

        var method = MissingOperationAnnotation.class.getDeclaredMethods()[0];
        var e = assertThrows(NullPointerException.class, () -> new MethodBinder(method));
        assertTrue(e.getMessage().contains("must be annotated"));
    }
}

File: cqf-fhir-utility/src/test/java/org/opencds/cqf/fhir/utility/operation/MethodBinderTest.kt
package org.opencds.cqf.fhir.utility.operation

import ca.uhn.fhir.model.api.annotation.Description
import ca.uhn.fhir.rest.annotation.IdParam
import ca.uhn.fhir.rest.annotation.Operation
import ca.uhn.fhir.rest.annotation.OperationParam
import java.lang.reflect.Method
import org.hl7.fhir.dstu3.model.Measure
import org.hl7.fhir.instance.model.api.IBaseParameters
import org.hl7.fhir.instance.model.api.IBaseResource
import org.hl7.fhir.r4.model.IdType
import org.hl7.fhir.r4.model.Parameters
import org.hl7.fhir.r4.model.StringType
import org.junit.Assert
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Test

@Suppress("unused", "UNUSED_PARAMETER")
class MethodBinderTest {
    // These are simply a bunch of example method signatures to test the MethodBinder class
    internal inner class ExampleMethods {
        fun noAnnotation() {}

        @Operation(name = "idFirst") fun idFirst(@IdParam id: IdType?) {}

        @Operation(name = "idLast")
        fun idLast(@OperationParam(name = "param") param: StringType?, @IdParam id: IdType?) {}

        @Operation(name = "noId") fun noId(@OperationParam(name = "param") param: StringType?) {}

        @Operation(name = "manyIds") fun manyIds(@IdParam id: IdType?, @IdParam id2: IdType?) {}

        @Operation(name = "idIsNotIdType") fun idIsNotIdType(@IdParam id: StringType?) {}

        @Operation(name = "extraFirst")
        fun extraFirst(
            @ExtraParams extras: IBaseParameters?,
            @OperationParam(name = "param") param: String?,
        ) {}

        @Operation(name = "extraLast")
        fun extraLast(
            @OperationParam(name = "param") param: StringType?,
            @ExtraParams parameters: IBaseParameters?,
        ) {}

        @Operation(name = "manyExtra")
        fun manyExtra(
            @ExtraParams parameters: IBaseParameters?,
            @ExtraParams parameters2: IBaseParameters?,
        ) {}

        @Operation(name = "extraIsNotParametersType")
        fun extraIsNotParametersType(@ExtraParams idType: IdType?) {}

        @Operation(name = "operationParamIsNotFhirType")
        fun operationParamIsNotFhirType(@OperationParam(name = "stringType") stringType: String?) {}

        @Description("has a description")
        @Operation(name = "hasDescription")
        fun hasDescription(@IdParam id: IdType?) {}

        @Operation(name = "hasCanonical", canonicalUrl = "http://example.com")
        fun hasCanonical(@IdParam id: IdType?) {}

        @Operation(name = "hasTypeName", typeName = "Measure")
        fun hasTypeName(@OperationParam(name = "stringType") stringType: StringType?) {}

        @Operation(name = "hasType", type = Measure::class)
        fun hasType(@OperationParam(name = "stringType") stringType: StringType?) {}

        @Operation(name = "noArgs") fun noArgs() {}

        @Operation(name = "\$nameNormalizes") fun nameNormalizes() {}
    }

    @Test
    fun noAnnotation_throws() {
        assertThrows(IllegalArgumentException::class.java) {
            MethodBinder(methodByName("noAnnotation"))
        }
    }

    @Test
    fun idFirst() {
        val m = methodBinderByName("idFirst")
        assertEquals(Scope.INSTANCE, m.scope)
        assertEquals("idFirst", m.name)
        assertEquals(1, m.parameters.size.toLong())
    }

    @Test
    fun idLast_throws() {
        assertThrows(IllegalArgumentException::class.java) { methodBinderByName("idLast") }
    }

    @Test
    fun manyIds_throws() {
        assertThrows(IllegalArgumentException::class.java) { methodBinderByName("manyIds") }
    }

    @Test
    fun idParamNotIdType_throws() {
        assertThrows(IllegalArgumentException::class.java) { methodBinderByName("idIsNotIdType") }
    }

    @Test
    fun noId() {
        val m = methodBinderByName("noId")
        assertEquals(Scope.SERVER, m.scope)
        assertEquals("noId", m.name)
        assertEquals(1, m.parameters.size.toLong())
    }

    @Test
    fun extraFirst_throws() {
        assertThrows(IllegalArgumentException::class.java) { methodBinderByName("extraFirst") }
    }

    @Test
    fun extraLast() {
        val m = methodBinderByName("extraLast")
        assertEquals(Scope.SERVER, m.scope)
        assertEquals("extraLast", m.name)
        assertEquals(2, m.parameters.size.toLong())
    }

    @Test
    fun manyExtra_throws() {
        assertThrows(IllegalArgumentException::class.java) { methodBinderByName("manyExtra") }
    }

    @Test
    fun extraIsNotParametersType_throws() {
        assertThrows(IllegalArgumentException::class.java) {
            methodBinderByName("extraIsNotParametersType")
        }
    }

    @Test
    fun operationParamIsNotFhirType_throws() {
        assertThrows(IllegalArgumentException::class.java) {
            methodBinderByName("operationParamIsNotFhirType")
        }
    }

    @Test
    fun binderGetsDescription() {
        val m = methodBinderByName("hasDescription")
        assertEquals("has a description", m.description.value)
    }

    @Test
    fun binderGetsCanonical() {
        val m = methodBinderByName("hasCanonical")
        assertEquals("http://example.com", m.canonicalUrl)
    }

    @Test
    fun binderGetsMethod() {
        val m = methodBinderByName("idFirst")
        Assert.assertNotNull(m.method)
        assertEquals(m.method.name, "idFirst")
    }

    @Test
    fun binderGetsOperationAnnotation() {
        val m = methodBinderByName("idFirst")
        Assert.assertNotNull(m.operation)
        assertEquals(m.operation.name, "idFirst")
    }

    @Test
    fun hasType() {
        val m = methodBinderByName("hasType")
        assertEquals("Measure", m.typeName)
        assertEquals(Scope.TYPE, m.scope)
    }

    @Test
    fun hasTypeName() {
        val m = methodBinderByName("hasTypeName")
        assertEquals("Measure", m.typeName)
        assertEquals(Scope.TYPE, m.scope)
    }

    @Test
    fun noArgs() {
        val m = methodBinderByName("noArgs")
        assertEquals(Scope.SERVER, m.scope)
    }

    @Test
    fun nameNormalizes() {
        val m = methodBinderByName("nameNormalizes")
        assertEquals("nameNormalizes", m.name)
    }

    @Test
    fun unboundArgs_throws() {
        val m = methodBinderByName("noArgs")

        val provider = ExampleMethods()
        val parameters = Parameters().addParameter("anyParam", StringType("anyValue"))
        val e =
            assertThrows(IllegalArgumentException::class.java) {
                m.bind(provider, null, parameters)
            }
        Assert.assertTrue(e.message!!.contains("not bound"))
    }

    @Test
    fun idWithoutInstanceScope_throws() {
        val m = methodBinderByName("noArgs")

        val provider = ExampleMethods()
        val id = IdType("Library/123")
        val e = assertThrows(IllegalArgumentException::class.java) { m.bind(provider, id, null) }
        Assert.assertTrue(e.message!!.contains("non-instance"))
    }

    @Test
    fun instanceScopeWithoutId_throws() {
        val m = methodBinderByName("idFirst")

        val provider = ExampleMethods()
        val e = assertThrows(IllegalArgumentException::class.java) { m.bind(provider, null, null) }
        Assert.assertTrue(e.message!!.contains("id required"))
    }

    @Test
    fun missingOperationParamAnnotation_throws() {
        class MissingOperationAnnotation {
            @Operation(name = "missingOperationParam")
            fun missingOperationParam(param: StringType?): IBaseResource? {
                return null
            }
        }

        val method = MissingOperationAnnotation::class.java.declaredMethods[0]
        val e = assertThrows(IllegalArgumentException::class.java) { MethodBinder(method) }
        Assert.assertTrue(e.message!!.contains("must be annotated"))
    }

    @Test
    fun conflictingAnnotations_throws() {
        class ConflictingAnnotation {
            @Operation(name = "function")
            fun function(@IdParam @OperationParam(name = "id") id: IdType?): IBaseResource? {
                return null
            }
        }

        val method = ConflictingAnnotation::class.java.declaredMethods[0]
        val e = assertThrows(IllegalArgumentException::class.java) { MethodBinder(method) }
        Assert.assertTrue(e.message!!.contains("one of"))
    }

    @Test
    fun missingOperationAnnotation_throws() {
        class MissingOperationAnnotation {
            fun function(@IdParam id: IdType?): IBaseResource? {
                return null
            }
        }

        val method = MissingOperationAnnotation::class.java.declaredMethods[0]
        val e = assertThrows(IllegalArgumentException::class.java) { MethodBinder(method) }
        Assert.assertTrue(e.message!!.contains("must be annotated"))
    }

    companion object {
        private val methods: List<Method> =
            listOf(*ExampleMethods::class.java.declaredMethods)

        private fun methodByName(name: String): Method {
            return methods.stream().filter { x: Method -> x.name == name }.findFirst().get()
        }

        private fun methodBinderByName(name: String): MethodBinder {
            return MethodBinder(methodByName(name))
        }
    }
}
